### ISSUE TITLE
Performance Issue. Redundant usage of npm packages lodash and promise

### ISSUE THREAD

**Issue by pganshirt** [Demandware/sitegenesis#52]
## Problem ##
Usage of these modules causing performance degradation on client side due: 
1. Extra amount of code included to js bundle (uncompressed size 600 KB, without modules 140 KB); 
2. Code duplication from existing tools and libs. Some of the used methods could work faster using native or jQuery alternatives

Across 47 storefront JS
**promise** module used 3 times 
**lodash** module used 7 times

## Solution ##
Here is a listing of each used method and details how to avoid its usage

### lodash ###

* *_.each  -> jQuery.each* No major reason to use lodash (jQuery 2% faster http://jsperf.com/jquery-each-vs-underscore-each-vs-for-loops/4). 
* *_.isString -> typeof()==="string"* It is better to use native method. Nice wrapper could be moved to utils module. (Native 100% faster https://jsperf.com/isstring-vs-typeof-string/2)
* *_.merge -> jQuery.extend* Replace with its jQuery alternative (jQuery 26% faster http://jsperf.com/lodash-merge-vs-jquery-extend)
* *_().findIndex* used 1 time only so it could be replaced with very simple native cycle
* *_().contains -> jQuery.inArray* Could be replaced with jQuery alternative, no high performance operations usage detected (jQuery 10% slower https://jsperf.com/jquery-inarray-vs-underscore-contains)
* *_.map -> jQuery.map() or Native* jQuery is also slower here, however there is no big collections used to iterate (jQuery 3-times slower https://jsperf.com/native-vs-array-js-vs-underscore/174)

**NOTE:** If it is really needed - only lodash parts could be included to avoid extra code. So for example you could put something like this:

```
#!javascript

var map = require('lodash.map');
```

### promise ###
jQuery already has great mechanism for defered objects, so usage extra module is pointless
 
* *promise.resolve() -> https://api.jquery.com/category/deferred-object/*
* *promise.all() -> http://stackoverflow.com/questions/4878887/how-do-you-work-with-an-array-of-jquery-deferreds*

**Comment by pganshirt** [comment]
CreatedBy=@tnguyen14
I agree with the overall thesis that we could work on reducing the size of the bundle (though there are probably better way to reduce page load by 400Kb), though I do not want that reason to be an argument against using a library.

Here are my thoughts on the topics above:

- lodash - we could definitely use individual lodash packages like you suggest. There's no reason to use the entire lodash library. The reason it was included in the first place was that we felt as a reference application, we should include a useful tool such as lodash. In specific implementation, if bundle size is a concerned, it could be replaced then.

- jquery deferreds - I am not a fan of jQuery deferred because I think its concept (introduced in the early days of promises in JS) overly complicates promises. ES2015 spec for Promises is a lot simpler. So I'd rather use a polyfill or babel to have Promise than use jQuery deferred object. Also, there was a while when jQuery promise is not A+ compliant. I also find the jQuery syntax of `{success: function () {}, error: function () {}}` to be an anti-pattern.

Things like `.map` should be used according to ES2015 spec as well, and then polyfilled with babel, rather than using either jQuery or lodash.

**Comment by pganshirt** [comment]
CreatedBy=@spublika
Hi @tnguyen14,

I appreciate usage of these great libraries, but architectural speaking "We have to use as much as possible from existing libraries as well as from platform API." This is must be consistent especially talking Reference Application.

*lodash*

From the RefApp standpoint you have to show best practice minimizing JS amount and using existing methods instead of adding extra lib.
The rule here is simple: whatever could be re-used from existing libraries must be reused (unless it has some major issues). In examples I've seen in js code, there is no reason to use something different from what I have listed. Probably most useful usage here it is maps. So your solution with polyfill might be much better.

*promises*

A+ compliance shouldn't be an argument. And also I don't think it will take more effort once ES2015 will be launched in browsers. 
1. It is not of so much different. See PR, I've created https://bitbucket.org/demandware/sitegenesis/pull-requests/695/removing-redundant-promise-library-issue/diff
2. It will be A+ compliant soon  https://github.com/jquery/jquery/issues/1722
3. jQuery is getting better. And if we want to use it in future it is right thing to use it as much as possible.

The issue right now it is hard to read code that combined from jQuery deffereds and Promises like you've used for some of the AJAX calls. $.ajax is already promise.

As for Syntax - just don't use that syntax. :) It is easy to describe in styleguide.


For now my biggest concern it's actually promises. We are in progress getting rid of lodash. It more then 3 years we are getting rid of different stuff in SG. Lucky we can change something here.

Thanks.

**Comment by pganshirt** [comment]
CreatedBy=@tnguyen14
Why "A+ compliance shouldn't be an argument"? I think that is a biggest issue for me. We were exposing promises that are A+ compliant. Changing them to jQuery now would be breaking that API. I'd rather we wait until they become compliant if you're really adamant about using them.

If we are serious about making things ES2015-compliant, then I'd suggest we start using native promises now and use babel es2015-preset and babel-polyfill to write standard JS today. I don't see much of the point converting to jQuery deferred, only to convert them back to standard promises later on.

**Comment by pganshirt** [comment]
CreatedBy=@spublika
If it would be a real argument you would start using bable from the beginning to keep it consistent with one library supporting.

With babelify it makes much more sense. But rules still need to be defined and documented. If Demanware will take it as a standard we will make sure it is standard in our company as well. 

We could install babellify it works perfectly with browserify and then change "includeExternals" to false, making this much more streamlined.

What do you think about it?

And second question are you serious to be ES2015-compliant (I'm asking because you mentioned it)? It will make this world much cooler, but needs to be reasonably integrated!

**Comment by pganshirt** [comment]
CreatedBy=@tnguyen14
I agree about babel and babelify. It would be a very nice flow.

About ES2015 - I think we should definitely discuss that as a project. I think that there are definite gains using standard APIs today. We might not want to jump on every ES2015 features, but some of them will make our jobs a lot easier.

**Comment by pganshirt** [comment]
CreatedBy=@spublika
That sounds great!
So can we now exclude Promise lib and optimize lodash usage for now? So it will be not mixed up with everything else.

**Comment by pganshirt** [comment]
CreatedBy=@tnguyen14
Based on our discussion on Slack, I think we can start using individual lodash methods instead of the whole library, i.e. `var map = require('lodash/map');`.

I'd either wait on the promise refactor until we use babel, or use a promise polyfill, such as https://github.com/jakearchibald/es6-promise or babel-polyfill.

**Comment by pganshirt** [comment]
CreatedBy=@jenssimon
Don't use the modularized Lodash. Don't use Lodash at all. SG with Lodash modularized brings it up to ~508kB. After switching to Underscore the file size of the bundle is ~452kB. And there's enough space for easy optimizations. Using jQuery UI as modules shrinks down the size to ~300kB. And if we can get rid off all those stupid bloated libraries and jQuery Plugins (Lodash/Underscore, jQuery UI, jQuery Validation, jCarousel, ...) we can reach file sizes of ~200kB (and it's fun!!!). And when you use libraries to polyfill things like Promise then don't let users with modern browsers pay for it. Polyfills can be loaded conditionally. Also stop using Browserify and switch to Webpack (all of the mentioned solutions are easy to achieve, but not with Browserify).

**Comment by pganshirt** [comment]
CreatedBy=@SqrTT
Agree with @tnguyen14, we use in project individual methods such as `var map = require('lodash/map');` same is valid for jQuery UI, eq. `require('jquery-ui/dialog')`, `require('jquery-ui/tooltip')` and Browserify will bundle only needed modules and it's dependencies. And overall js file size is ~120kB gzipped.

**Comment by pganshirt** [comment]
CreatedBy=@jenssimon
When we talk about gzipped file sizes I get ~91kB (~300 kB uncompressed, only minified), Also includes the stuff not bundled via Browserify in SG (jQuery, jQuery plugins, ...)

**Comment by pganshirt** [comment]
CreatedBy=@tnguyen14
@jenssimon I've been wondering about webpack myself. I think it's interesting, but feels a bit too complex for SG. If you'd like to put something together in the form of a PR for how a webpack set up for SG would look like, I'd love to take a look?
